cargo-pgrx/src/command/upgrade.rs

```diff
@@ -0,0 +1,216 @@
+            cargo_edit::Source::Git(git) => git.version.as_ref(),
+            cargo_edit::Source::Workspace(_ws) => {
+                error!(
+                    "Cannot fetch the version of \
+                    a package because it is set in the \
+                    workspace, please run `cargo pgrx \
+                    upgrade` in the workspace directory."
+                );
+                None
+            }
+        }
+    }
+}
+fn replace_version<S>(
+    new_version: &str,
+    crate_root: &Path,
+    key: &mut toml_edit::KeyMut,
+    dep: &mut toml_edit::Item,
+    mut parsed_dep: Dependency,
+    mut source: S,
+) -> eyre::Result<()>
+where
+    S: Clone + DependencySource,
+    cargo_edit::Source: From<S>,
+{
+    let dep_name = key.get();
+    let ver_maybe = source.get_version();
+    match ver_maybe {
+        Some(v) => {
+            debug!("{dep_name} version is {v}")
+        }
+        None => return Err(eyre!("No version field for {dep_name}, cannot upgrade.")),
+    }
+    source.set_version(new_version);
+    parsed_dep = parsed_dep.set_source(source);
+
+    parsed_dep.update_toml(crate_root, key, dep);
+    Ok(())
+}
+impl Upgrade {
+    fn update_dep(
+        &self,
+        path: &PathBuf,
+        mut key: KeyMut,
+        dep: &mut toml_edit::Item,
+    ) -> eyre::Result<()> {
+        let dep_name_string = key.get().to_string();
+        let dep_name = dep_name_string.as_str();
+        let parsed_dep: Dependency = match Dependency::from_toml(path.as_path(), dep_name, dep) {
+            Ok(dependency) => dependency,
+            Err(e) => {
+                return Err(eyre!(
+                    "Could not parse dependency \
+                entry for {dep_name} due to error: {e}"
+                ))
+            }
+        };
+        let reg_url = registry_url(path, parsed_dep.registry())
+            .map_err(|e| eyre!("Unable to fetch registry URL for path: {e}"))?;
+        let target_version = match self.target_version {
+            Some(ref ver) => Some(ver.clone()),
+            None => cargo_edit::get_latest_dependency(
+                dep_name,
+                self.allow_prerelease,
+                None,
+                path,
+                Some(&reg_url),
+            )
+            .map_err(|e| {
+                eyre!(
+                    "Unable to fetch the latest version \
+                        for crate {dep_name} due to {e}"
+                )
+            })?
+            .version()
+            .map(|s| s.to_string()),
+        };
+        let target_version = match target_version {
+            Some(ver) => ver,
+            None => {
+                return Err(eyre!(
+                    "Unable to update {dep_name} \
+                , no provided crate version and a latest version \
+                could not be retrieved from the registry."
+                ))
+            }
+        };
+
+        if let Some(source) = parsed_dep.source().cloned() {
```

Comment:
hmm, when is it possible for the source to be `None`?

Reply:
This is implemented within `cargo-edit`, it looks like it initializes as `None` and is filled in later as the manifest file is parsed. Reading `cargo-edit`'s source code at https://github.com/killercup/cargo-edit/blob/master/src/dependency.rs#L269, it looks like `source` will always be `Some()` if the `Dependency` struct is instantiated by `from_toml()` This might just be an API design mistake on their part, or there might be some deeper reason I cannot spot here.

Reply:
Resolved in d0f202cb